woocommerce/woocommerce-gutenberg-products-block
# Product Elements: Use button colors as default for Sale Badge block

In this PR, the `wp_theme_json_data_theme` filter is utilized to set default background and text colors for the Sale Badge block. The following logic is applied:
1. If the theme has a `theme.json` file and defines colors for the button element, use those colors for the Sale Badge block.
2. If not, fallback to using colors from [style.scss](https://github.com/woocommerce/woocommerce-blocks/blob/ab48e8cfae5ed9abd89e0cd04c3d59ccb0f34e62/assets/js/atomic/blocks/product-elements/sale-badge/style.scss#L10-L11)

### Walkthrough 
* Add a new method `add_default_sale_badge_styles` to the `ProductSaleBadge` class, which filters the theme data and adds default styles for the Sale Badge block ([link](https://github.com/woocommerce/woocommerce-blocks/pull/9068/files?diff=unified&w=0#diff-2f63a5e1689a2f8642d3a974cc857d1bc8d037abc3efe1ee4ccad08fe5381b5cR67-R128))
* Register the `add_default_sale_badge_styles` method as a callback for the `woocommerce_blocks_get_theme_data` filter in the `register_block_type` method of the same class ([link](https://github.com/woocommerce/woocommerce-blocks/pull/9068/files?diff=unified&w=0#diff-2f63a5e1689a2f8642d3a974cc857d1bc8d037abc3efe1ee4ccad08fe5381b5cR67-R128))

Fixes #8151 

### Testing

#### Testing with different themes
1. Create a new post
2. Add products block to the post. 
3. Add the "On-Sale Badge" as an inner block
<img width="200" alt="image" src="https://user-images.githubusercontent.com/16707866/232464386-1fea6565-b607-4cee-89c3-f2c49b007f74.png">
4. Activate TT3 and verify that the Sale Badge inherits button element colors on both the Editor and Frontend
<img width="225" alt="image" src="https://user-images.githubusercontent.com/16707866/232464908-521ecf40-c1f7-4c25-a46f-32912e05bef0.png">
5. Activate TT2; this theme has a theme.json file but does not define button element colors. Verify that the Sale Badge colors fallback to the colors defined in style.scss on both the Frontend and Editor
<img width="228" alt="image" src="https://user-images.githubusercontent.com/16707866/232465480-843b9526-cbe2-4890-b050-c86d2ac16dd9.png">
6. Activate TT1, which does not have a theme.json file. Verify that the Sale Badge colors fallback to the colors defined in style.scss. Repeat the same for the Storefront theme.
<img width="257" alt="image" src="https://user-images.githubusercontent.com/16707866/232465775-5236350a-4510-4f13-bfa3-9618402205ef.png">

In summary, ensure that the Sale Badge uses button element colors from the theme.json file when available. Otherwise, fallback to the colors defined in style.scss.

#### Test using Global styles UI
1. Activate the TT3 theme or another block theme
2. By default, the Sale Badge should inherit button element colors
3. Open the Global Styles sidebar by clicking the "Styles" button at the top right corner of the editor.

https://user-images.githubusercontent.com/16707866/228781997-24a9a29f-4be7-46d4-b1b3-41520692d86d.mov

4. In the Global Styles sidebar, locate the "On-Sale Badge" block
5. Change the background and text color, and Save.
6. Verify that the changes are reflected correctly on both the Frontend and Editor


### WooCommerce Visibility




### Changelog

> On-Sale Badge: Use background and text color of the button element if not defined in theme.json
